src/commands/publish/script_upload_form.rs

```diff
@@ -0,0 +1,112 @@
+use log::info;
+
+use reqwest::multipart::{Form, Part};
+use std::fs;
+use std::path::Path;
+
+use crate::commands::build::wranglerjs::Bundle;
+use crate::settings::project::{Project, ProjectType};
+
+use super::{krate, Package};
+
+pub fn build_script_upload_form(project: &Project) -> Result<Form, failure::Error> {
+    let project_type = &project.project_type;
+    let script_upload_form = match project_type {
+        ProjectType::Rust => {
+            info!("Rust project detected. Publishing...");
+            build_multipart_script()?
+        }
+        ProjectType::JavaScript => {
+            info!("JavaScript project detected. Publishing...");
+            build_js_script()?
+        }
+        ProjectType::Webpack => {
+            info!("Webpack project detected. Publishing...");
+            build_webpack_form()?
+        }
+    };
+
+    Ok(script_upload_form)
+}
+
+fn build_js_script() -> Result<Form, failure::Error> {
+    let package = Package::new("./")?;
+    let script_path = package.main()?;
+    let metadata_json = r#"{"body_part":"script","bindings":[]}"#;
+
+    let metadata = Part::text(metadata_json)
+        .file_name("metadata.json")
+        .mime_str("application/json")?;
+
+    Ok(Form::new()
+        .file("script", &script_path)
+        .unwrap_or_else(|_| panic!("{} not found. Did you rename your js files?", &script_path))
+        .part("metadata", metadata))
+}
+
+fn build_multipart_script() -> Result<Form, failure::Error> {
+    let name = krate::Krate::new("./")?.name.replace("-", "_");
+    build_generated_dir()?;
+    concat_js(&name)?;
+
+    let metadata_path = "./worker/metadata_wasm.json";
+    let wasm_path = &format!("./pkg/{}_bg.wasm", name);
+    let script_path = "./worker/generated/script.js";
+
+    Ok(Form::new()
+        .file("metadata", metadata_path)
+        .unwrap_or_else(|_| panic!("{} not found. Did you delete it?", metadata_path))
+        .file("wasmprogram", wasm_path)
+        .unwrap_or_else(|_| panic!("{} not found. Have you run wrangler build?", wasm_path))
+        .file("script", script_path)
+        .unwrap_or_else(|_| panic!("{} not found. Did you rename your js files?", script_path)))
+}
+
+fn build_generated_dir() -> Result<(), failure::Error> {
+    let dir = "./worker/generated";
+    if !Path::new(dir).is_dir() {
+        fs::create_dir("./worker/generated")?;
+    }
+
+    Ok(())
+}
+
+fn concat_js(name: &str) -> Result<(), failure::Error> {
+    let bindgen_js_path = format!("./pkg/{}.js", name);
+    let bindgen_js: String = fs::read_to_string(bindgen_js_path)?.parse()?;
+
+    let worker_js: String = fs::read_to_string("./worker/worker.js")?.parse()?;
+    let js = format!("{} {}", bindgen_js, worker_js);
+
+    fs::write("./worker/generated/script.js", js.as_bytes())?;
+    Ok(())
+}
+
+fn build_webpack_form() -> Result<Form, failure::Error> {
```

Comment:
With the WorkerBundle/Worker abstraction I was able to generate the same form for every backend. All have `script` and accept a `bindinds` array (can be empty).
All the specific paths were configured by each backend.

Reply:
that's my next pr 😄